### ISSUE TITLE
🔴 Fix error handling when creating a request fails

### ISSUE THREAD

**Issue by neil-marcellini** [Expensify/App#22126]
# Problem
If there is an error creating a manual request we should show a RBR leading to the error message and it should be able to be dismissed, and everything should get cleaned up. See the [report in Slack](https://expensify.slack.com/archives/C05DWUDHVK7/p1688154537120499) for reproduction steps and a video.

# Why is this important?
We want the user to have a good experience even when something goes wrong unexpectedly. Also, for distance requests we want to build on top of the error handling for manual requests, so it should work before we start building on top of it.

# Solution
Fix error handling when creating a request fails

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0133f67d73d5f102fa</li>
        <li>Upwork Job ID: 1675927782156615680</li>
        <li>Last Price Increase: 2023-07-03</li>
    </ul>
</details>

**Comment by neil-marcellini** [comment]
I'll try to get to this tomorrow.

**Comment by neil-marcellini** [comment]
I still haven't had time for this, but tomorrow is looking good.

**Comment by neil-marcellini** [comment]
I made some good progress today. I got the RBR showing the path to the failed request action, I disabled the pay button on the report, and I displayed the error message under the request action.

Next time I need to delete all the optimistic data when the error message is dismissed. I also need to figure out why the createChat error is getting cleared for the iouReport.

**Comment by neil-marcellini** [comment]
Today I worked on cleaning up the optimistic data for the failed request when the message is dismissed. I'm getting close with that, but there are still a handful of problems, such as the error message getting displayed twice.

**Comment by neil-marcellini** [comment]
I'm going to be OOO tomorrow and Friday and I can pick this up when I get back. I'll also post to engineering chat to see if anyone wants to take it over while I'm out.

**Comment by neil-marcellini** [comment]
I didn't get to this yesterday but I'll try to fit it in today.

**Comment by neil-marcellini** [comment]
Still haven't had time to get back to this.

**Comment by neil-marcellini** [comment]
I'm downgrading this to a weekly too.

**Comment by neil-marcellini** [comment]
Still haven't worked on it anymore but I'm hoping to squeeze it in soon

**Comment by neil-marcellini** [comment]
No time with distance requests, but this is related so I should probably get back to it soon.

**Comment by neil-marcellini** [comment]
I finished all the distance request issues I was assigned to, so I'm going to bump up the priority on this one and take it on.

**Comment by neil-marcellini** [comment]
I merged main into the PR, did a little bit of clean up, and removed one of the duplicate error messages. I'm still working on properly cleaning up the optimistic data from when the money request was created.

**Comment by neil-marcellini** [comment]
Yesterday we decided to focus on CRITICAL and HIGH priority issues only so I jumped in to help there instead.

**Comment by neil-marcellini** [comment]
Moving it back to a weekly due to its lower priority in the distance project

**Comment by neil-marcellini** [comment]
I'm working on this again 👀

**Comment by neil-marcellini** [comment]
I'm encountering some strange behavior where it seems like I can't delete a report action with Onyx merge or even set.

**Comment by thienlnam** [comment]
Gonna push this forward while Neil is OOO

**Comment by neil-marcellini** [comment]
I'm currently trying to fix this bug on my [draft PR](https://github.com/Expensify/App/pull/22725).
1. Sign in
2. Request money and have it fail
3. Open the report
4. Dismiss the error message
5. Refresh the page

A blank action shows up with an error message.

**Comment by neil-marcellini** [comment]
For some reason when I get all the report actions on the report, delete the failed IOU report action, and then call Onyx.set with the action removed, it doesn't work.

However when I do this on a test collection key it does work. I'm very confused.

**Comment by thienlnam** [comment]
Can you link to where you call the Onyx.set? I wonder if this is due to the null value setting bug in Onyx

**Comment by neil-marcellini** [comment]
Woohoo Jack and Tim helped me figure it out. Merging null in Onyx doesn't delete entries right now, nested or not. Onyx.set works but only if you clone the data before deleting and entry and passing that data to Onyx.set.

**Comment by neil-marcellini** [comment]
Not overdue

**Comment by thienlnam** [comment]
We should probably figure out what we want to do with this issue - we verified that when the transaction fails the error is handled successfully. The only case that isn’t handled is if the IOU fails to created, but that’s not isolated to distance requests

https://expensify.slack.com/archives/C05DWUDHVK7/p1696373137587939

**Comment by neil-marcellini** [comment]
I made some progress. Now a failed expense report will have its preview deleted when it's opened.

**Comment by neil-marcellini** [comment]
I put the PR up for review. Tim requested some small changes and also found quite a few bugs. I took a step back and thought about what I've implemented so far vs what we ultimately want  for handling errors on distance requests, and I think my current PR is almost counter productive unfortunately. Here are a few problems.

- the RBR should direct to the lowest level context item (the individual request), vs the top level (workspace chat). I will makes sure there's broad agreement on this point
- similarly, it should always be possible to view the individual failed request
- a failed money request should be shown on an existing expense report when there is one, but currently a new failed expense report is shown
- it should be possible to fix a failed distance request as described in [this manual test case](https://docs.google.com/document/d/16Q9sE-991CZ4wUcalOs_U-qyEEza_C9tZwj6G5lHUp0/edit#heading=h.45t4k5tpvn7w)

With those problems in mind I think it would be better to close this issue and focus instead on [Handle errors with geocoding for offline distance requests](https://github.com/Expensify/App/issues/22719). I will likely keep working on the same front end PR but instead of testing an artificially failing manual request I'll use a test case that can actually happen, which is errors when geocoding an offline distance request. Another reason to test with the actual flow is that I might need to make backend changes specific to when distance requests fail, such as failing the distance request but successfully creating the expense report.
